### PR TITLE
Rebase stm32.js

### PR DESCRIPTION
**Pull request by haslinghuis**
Rebase of stm32.js:

- [x] fix lexical scoping
- [x] change strings into template literals
- [x] remove type coercion
- [x] change (most) variables from kebab_case to camelCase 

Only one file to go (usbstm32dfu.js)

### CODE REVIEW

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> FAIL
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> PASS
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> PASS
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Review by Asizon — APPROVED** [review]

**Review by McGiverGim — APPROVED** [review]
Good cleaning! I have not tested it, but seems good to me...

**Review by ctzsnooze — APPROVED** [review]

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> PASS
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> PASS
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

File: src/js/protocols/stm32.js [hunk 1]
```diff
@@ -265,69 +265,71 @@ STM32_protocol.prototype.initialize = function () {
...
+    this.bytesToRead = bytesToRead;
     this.read_callback = callback;
 
     // empty receive buffer before next command is out
     this.receive_buffer = [];
 
     // send over the actual data
-    serial.send(bufferOut, function (writeInfo) {});
+    serial.send(bufferOut, function (writeInfo) {
+        //
```
Comment (mikeller) [inline src/js/protocols/stm32.js:322]: Looking at `src/js/serial.js`, `callback` seems to be optional, so we can probably just omit this.

File: src/js/protocols/stm32.js [hunk 2]
```diff
@@ -422,31 +423,31 @@ STM32_protocol.prototype.verify_chip_signature = function (signature) {
 // second_array = usually verify_hex array
 // result = true/false
 STM32_protocol.prototype.verify_flash = function (first_array, second_array) {
-    for (var i = 0; i < first_array.length; i++) {
+    for (let i = 0; i < first_array.length; i++) {
```
Comment (mikeller) [inline src/js/protocols/stm32.js:426]: Did you want to fix `first_array` as well?

File: src/js/protocols/stm32.js [hunk 3]
```diff
@@ -265,69 +265,71 @@ STM32_protocol.prototype.initialize = function () {
...
+        this.read_callback(fetched);
     }
 };
 
 // we should always try to consume all "proper" available data while using retrieve
 STM32_protocol.prototype.retrieve = function (n_bytes, callback) {
     if (this.receive_buffer.length >= n_bytes) {
         // data that we need are there, process immediately
-        var data = this.receive_buffer.slice(0, n_bytes);
+        const data = this.receive_buffer.slice(0, n_bytes);
```
Comment (mikeller) [inline src/js/protocols/stm32.js:289]: Did you want to fix `n_bytes` as well?

File: src/js/protocols/stm32.js [hunk 4]
```diff
@@ -631,49 +635,50 @@ STM32_protocol.prototype.upload_procedure = function (step) {
...
-
-                                    var checksum = array_out[0];
-                                    for (var i = 0; i < bytes_to_write; i++) {
-                                        array_out[i + 1] = self.hex.data[flashing_block].data[bytes_flashed]; // + 1 because of the first byte offset
+                            const addressArray = [(address >> 24), (address >> 16), (address >> 8), address];
+                            const addressChecksum = addressArray[0] ^ addressArray[1] ^ addressArray[2] ^ addressArray[3];
+                            // write start address + checksum
+                            self.send([addressArray[0], addressArray[1], addressArray[2], addressArray[3], addressChecksum], 1, function (_reply) {
+                                if (self.verify_response(self.status.ACK, _reply)) {
+                                    const ArrayOut = Array.from(bytesToWrite + 2); // 2 byte overhead [N, ...., checksum]
```
Comment (mikeller) [inline src/js/protocols/stm32.js:665]: Not sure why this is capitalised?

File: src/js/protocols/stm32.js (hunk 1 above)
Comment (haslinghuis) [reply src/js/protocols/stm32.js]: Reduced.

File: src/js/protocols/stm32.js (hunk 2 above)
Comment (haslinghuis) [reply src/js/protocols/stm32.js]: Renamed

File: src/js/protocols/stm32.js (hunk 3 above)
Comment (haslinghuis) [reply src/js/protocols/stm32.js]: Renamed

File: src/js/protocols/stm32.js (hunk 4 above)
Comment (haslinghuis) [reply src/js/protocols/stm32.js]: Yes, as original, fixed now.

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> PASS
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Comment by haslinghuis** [comment]
Do I need to do anything here to get it approved?

**Review by McGiverGim — APPROVED** [review]

**Review by blckmn — APPROVED** [review]

**Review by limonspb — APPROVED** [review]

**Comment by blckmn** [comment]
AUTOMERGE: (PASS)
 - github identifies PR as mergeable -> PASS
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> PASS
 - assigned to an approver -> PASS
 - approver count at least three -> PASS

**Pull request merged**